features/FxA-45-device-registration-api/FX-DESKTOP.md

```diff
@@ -0,0 +1,122 @@
+
+## Existing profile, no FxA deviceID, has FxA sessionToken
+
+This is a user who has previously signed in to FxA to connect
+to Sync. This user will have an FxA sessionToken, but may not yet
+have an FxA deviceID.
+
+The user facing goal is for the device to appear in the user's device
+list without additional user action. To facilitate this, Firefox
+should register the device on behalf of the user.
+
+On startup, Firefox should attempt to register the device with the
+FxA auth server using the name and type currently stored in the
+user's Sync preferences.
+
+Firefox will send the following information to the FxA auth server to
+register the device:
+
+* deviceName (from Sync preferences)
+* deviceType (from Sync preferences)
+* pushCallback (from Push preferences)
+* pushPublicKey (from Push preferences)
+
+### Registration success
+
+The FxA auth server will respond with:
+
+* deviceID
+* deviceName
+* deviceType
+* pushCallback
+* pushPublicKey
+
+Firefox must then store the returned deviceID.
+
+### Registration failure
+
+If registration fails because of a network error, Firefox should re-attempt
+to register after a TBD interval has elapsed.
+
+If registration fails for other reasons - WHAT HERE?
+
+## Existing profile, has FxA deviceID, has FxA sessionToken
+
+This is a user who has previously signed in to FxA to
+connect to Sync and their device is already registered.
+
+This device has already sent its initial name to the FxA
+auth server, but the user may have remotely disconnected
+the device, remotely reset the account password, or remotely
+changed the account password. The FxA auth server is considered
+the canonical source of truth for device state.
+
+On startup, Firefox should fetch the current device list
+from the FxA auth server and compare internal Firefox state
+with the state returned from the auth server.
+
+### Device list fetch success
+
+The FxA auth server will respond with an array of devices with the
+following information:
+
+* deviceID
+* deviceName
+* deviceType
+* pushCallback
+* pushPublicKey
+
+If the user has remotely disconnected the device, Firefox
+should disconnect the user from Sync (and wipe their FxA sessionToken?)
+
+If the user has remotely changed or reset the account password,
+Firefox should ask the user to re-authenticate with Firefox accounts.
+
+### Device list fetch failure
+
+If device list fails because of a network error, Firefox should re-attempt
+to re-fetch the list after a TBD interval has elapsed.
+
+## Firefox receives an `fxaccounts:login` message
+
+A user has just authenticated with Firefox Accounts. The message payload
+will contain both a sessionToken.
+
+If Firefox already has an FxA deviceID, then the user re-authenticated
+for some reason (reset password, change password).
+
+XXX Will the devices API automatically re-associate the old deviceID with the
+new sessionToken when the user signs in? Is there any mechanism for the auth
+server to do that automatically?
```

Comment:
Yep.  I see three broad approaches:

1. Have the device pass it as a query parameter, e.g. `context=fx_desktop_v2&device=ABCDEF` and maybe even `device=` to signal that it has no id, but wants one.
2. Coordinate the deviceid as part of the device handshake.
3. Don't worry about it in the content-server at all, and just have the device bootstrap itself into getting a device-id once it has a session token.

The *eventual* right thing is probably (2), but I don't have strong feelings about whether (1) or (3) would work better for the initial version.

There are hypothetical reasons why the auth-server might like to know whether it's allocating a device-session or web-session token, e.g. https://github.com/mozilla/fxa-auth-server/issues/286, but YAGNI could easily overpower them.